### PR TITLE
[FLINK-40233][table] Correctly share the parsed JSON across `JSON_VALUE`/`JSON_QUERY` calls when the owning call is skipped because of `NULL` or not taken branch

### PR DESCRIPTION
**Pull request by snuyanzin**
## What is the purpose of the change

It might happen that `NULL` is cached while it should not for instance in case of 
```sql

SELECT json_value(v, CAST(NULL AS STRING)), json_value(v, '$.a')
FROM (values('{"a":1, "b":2}')) AS t(v);
-- returns (NULL, NULL), expected (NULL, 1)
```

## Brief change log

codegen

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

---

##### Was generative AI tooling used to co-author this PR?

- [ ] Yes (please specify the tool below)

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/codegen/JsonParseReuseTest.java [hunk 1]
```diff
@@ -133,9 +137,10 @@ void testJsonValueAndJsonQueryMixed() {
                 .containsExactlyInAnyOrder(
                         Row.of("account", "{\"city\":\"Munich\"}"),
                         Row.of("admin", "{\"city\":\"Berlin\"}"));
-        assertThat(countJsonParse(extractGeneratedCode(sql)))
+        final String code = extractGeneratedCode(sql);
+        assertThat(countJsonParse(code))
                 .as("JSON_VALUE + JSON_QUERY on the same input should parse once")
-                .isEqualTo(1);
+                .isOdd();
```
Comment (gustavodemorais) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/codegen/JsonParseReuseTest.java]: Leftover?

```suggestion
                .isOne();
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/codegen/JsonParseReuseTest.java [hunk 2]
```diff
@@ -149,9 +154,81 @@ void testThreeJsonFunctionCalls() {
...
-        assertThat(countJsonParse(extractGeneratedCode(sql)))
+        final String code = extractGeneratedCode(sql);
+        assertThat(countJsonParse(code))
                 .as("Three JSON function calls on the same input should parse once")
-                .isEqualTo(1);
+                .isOne();
+    }
+
+    @Test
+    void testReuseSurvivesCodeSplitting() {
```
Comment (gustavodemorais) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/codegen/JsonParseReuseTest.java:175]: Are you sure this is testing code splitting? 

extractGeneratedCode reads GeneratedClass.getCode(), which is the original source; splitting is applied in the constructor into the separate splitCode field and only that one is compiled (GeneratedClass.java:55-62, :97). The config only affects the collect(sql) half. I think it needs some tweaks in the test to test splitting

But to be honest is not the most important thing to test in the PR, so I'd not say this is a blocker for me and dropping the test would also be ok

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/codegen/JsonParseReuseTest.java [hunk 3]
```diff
@@ -165,4 +242,38 @@ void testDifferentJsonInputs() {
...
+        final List<Row> rows = new ArrayList<>();
+        bEnv.executeSql(sql).collect().forEachRemaining(rows::add);
+        assertThat(rows)
+                .containsExactlyInAnyOrder(
+                        Row.of(1, "account", "{\"city\":\"Munich\"}"),
+                        Row.of(2, "admin", "{\"city\":\"Berlin\"}"));
+    }
+
+    @Test
+    void testFilterAndProjectionShareParse() {
```
Comment (gustavodemorais) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/codegen/JsonParseReuseTest.java:307]: All tests use a plain column: could you add a test using a computed column like using TRIM(json_data) as the input to the json functions?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/codegen/JsonParseReuseTest.java [hunk 4]
```diff
@@ -165,4 +242,38 @@ void testDifferentJsonInputs() {
                 .as("JSON_VALUE calls on different inputs should parse separately")
                 .isEqualTo(2);
     }
+
```
Comment (gustavodemorais) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/codegen/JsonParseReuseTest.java:285]: I think I found one edge case, but it might have an easy solution

Repro
```
-- rows: (1000,'{"n":1}') (2000,'{"n":2}') (3000,'{"n":3}') (9000,'{"n":9}')
SELECT total FROM events MATCH_RECOGNIZE (
  ORDER BY rt
  MEASURES SUM(CAST(JSON_VALUE(A.f1, '$.n') AS INT)) AS total
  AFTER MATCH SKIP PAST LAST ROW
  PATTERN (A+ B)
  DEFINE A AS A.f0 < 9000, B AS B.f0 >= 9000
)
-- this branch: [3]  (= 1+1+1)
-- master:      [6]  (= 1+2+3)
```

The aggregate loops over every row of the match (MatchCodeGenerator.scala), but jsonParsed$0 = null; is a per-record statement, emitted once per processMatch - so rows 2..n reuse row 1's parse. Same cause in batch with table.exec.operator-fusion-codegen.enabled=true: CalcFusionCodegenSpec never emits reusePerRecordCode(), so the reset is dropped entirely.

Master is safe because the assignment was unconditional at the call site. The problem is that correctness now depends on which operator template wraps the expression. 

Possible easy solution: A local (addReusableLocalVariable, plus an explicit = null init) would fix both - reuseLocalVariableCode() is emitted in the fusion path and is keyed per method, so it resets inside the MATCH loop too

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/JsonParseReuse.scala [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      case None =>
+        val varName = CodeGenUtils.newName(ctx, "jsonParsed")
+        val methodName = CodeGenUtils.newName(ctx, "parseJson")
+        val typeName = classOf[SqlJsonUtils.JsonValueContext].getName
+        ctx.addReusableMember(s"$typeName $varName;")
+
+        // null means not parsed yet - jsonParse never returns null for a non-null input
+        ctx.addReusableMember(
+          s"""
+             |private $typeName $methodName(Object in) {
```
Comment (gustavodemorais) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/JsonParseReuse.scala]: Could we do this? 

```suggestion
             |private $typeName $methodName(${CodeGenUtils.BINARY_STRING} in) {
```

**Review by gustavodemorais — COMMENTED** [review]
Hey @snuyanzin, thanks quickly picking this one up. I've added some suggestions

**Review by gustavodemorais — APPROVED** [review]
LGTM. Thanks for fixing and addressing the comments @snuyanzin

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/JsonParseReuse.scala [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      case None =>
+        val varName = CodeGenUtils.newName(ctx, "jsonParsed")
+        val lastInputName = CodeGenUtils.newName(ctx, "jsonParsedInput")
+        val methodName = CodeGenUtils.newName(ctx, "parseJson")
+        val typeName = classOf[SqlJsonUtils.JsonValueContext].getName
+        val inputType = CodeGenUtils.BINARY_STRING
+        ctx.addReusableMember(s"$typeName $varName;")
+        ctx.addReusableMember(s"$inputType $lastInputName;")
+
+        // keyed on the immutable input so it re-parses on change; no per-record reset needed
```
Comment (wenshao) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/JsonParseReuse.scala:64]: **[Suggestion]** The comment claims the cache is safe because the input is "immutable," but `BinaryStringData` is technically mutable (it extends `LazyBinaryFormat` which has a public `setJavaObject` setter and package-private mutable fields). The actual safety invariant is narrower: `BinaryRowData.getString()` creates a **new** `BinaryStringData` object per call (via `fromAddress` → `new BinaryStringData(...)`), so reference identity changes per record — Concrete cost: a future maintainer optimizing row unboxing may read "immutable" and conclude the cache is safe regardless of reference identity, or may "fix" `!=` to `!in.equals()`, replacing the O(1) identity check with an O(n) byte comparison on every JSON call site.

```suggestion
      // Reference-equality cache: BinaryRowData.getString() creates a fresh
      // BinaryStringData per call, so a new record always brings a new reference.
      // Do NOT switch to .equals() — the O(1) identity check is the point.
```

_— qwen3.8-max-preview via Qwen Code /review (v0.21.2)_

**Review by wenshao — COMMENTED** [review]
Reviewed — no blockers. Suggestions are inline.

_— qwen3.8-max-preview via Qwen Code /review (v0.21.2)_
